# Lab | Data Cleaning and Formatting

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv


# Challenge 1: Data Cleaning and Formatting

## Exercise 1: Cleaning Column Names

To ensure consistency and ease of use, standardize the column names of the dataframe. Start by taking a first look at the dataframe and identifying any column names that need to be modified. Use appropriate naming conventions and make sure that column names are descriptive and informative.

*Hint*:
- *Column names should be in lower case*
- *White spaces in column names should be replaced by `_`*
- *`st` could be replaced for `state`*

In [190]:
import pandas as pd


In [191]:
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")


## Exercise 2: Cleaning invalid Values

The dataset contains columns with inconsistent and incorrect values that could affect the accuracy of our analysis. Therefore, we need to clean these columns to ensure that they only contain valid data.

Note that this exercise will focus only on cleaning inconsistent values and will not involve handling null values (NaN or None).

*Hint*:



In [192]:
df.head()

,Customer,ST,GENDER,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59%,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17%,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18%,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65%,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323


In [193]:
df.shape

(4008, 11)

In [194]:
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [195]:
df.columns

Index(['Customer', 'ST', 'GENDER', 'Education', 'Customer Lifetime Value',
       'Income', 'Monthly Premium Auto', 'Number of Open Complaints',
       'Policy Type', 'Vehicle Class', 'Total Claim Amount'],
      dtype='object')

In [196]:
df.isna().any()

Customer                     True
ST                           True
GENDER                       True
Education                    True
Customer Lifetime Value      True
Income                       True
Monthly Premium Auto         True
Number of Open Complaints    True
Policy Type                  True
Vehicle Class                True
Total Claim Amount           True
dtype: bool

In [197]:
df.isna().sum()

Customer                     2937
ST                           2937
GENDER                       3054
Education                    2937
Customer Lifetime Value      2940
Income                       2937
Monthly Premium Auto         2937
Number of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount           2937
dtype: int64

##Gender column contains various inconsistent values such as "F", "M", "Femal", "Male", "female", which need to be standardized, for example, to "M" and "F".*

In [198]:
df['GENDER'].value_counts()

GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

In [199]:
def standarized_gender(gender):
    if gender in ["Femal", "female", "F"]:
        return "F"
    elif gender in ["Male", "M"]:
        return "M"
    else:
        return gender

In [200]:
df['GENDER'] = df['GENDER'].apply(standarized_gender)

In [201]:
df.rename(columns={'GENDER': 'Gender'}, inplace=True)

In [202]:
df["Gender"].value_counts()

Gender
F    502
M    452
Name: count, dtype: int64

 *State abbreviations be can replaced with its full name, for example "AZ": "Arizona", "Cali": "California", "WA": "Washington"*

In [203]:
df['ST'].value_counts()

ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

In [204]:
city_changes = {"Oregon" : "Oregon" , "California" : "California" , "Arizona": "Arizona" , 'Cali': 'California', "Nevada": "Nevada", 
                "Washington" : "Washington",  'WA': 'Washington', 'AZ': 'Arizona'}
df['ST'] = df['ST'].replace(city_changes)

In [205]:
df.rename(columns={'ST': 'State'}, inplace=True)

*In education, "Bachelors" could be replaced by "Bachelor"*

In [206]:
df['Education'].value_counts()

Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

In [207]:
education_changes = { "Bachelors" : "Bachelor"}
df['Education'] = df['Education'].replace(education_changes)
df['Education']

0                     Master
1                   Bachelor
2                   Bachelor
3                   Bachelor
4       High School or Below
                ...         
4003                     NaN
4004                     NaN
4005                     NaN
4006                     NaN
4007                     NaN
Name: Education, Length: 4008, dtype: object

-#In Customer Lifetime Value, delete the `%` character*
- #In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [208]:
df["Customer Lifetime Value"]= df["Customer Lifetime Value"].str.rstrip("%")

-*In vehicle class, "Sports Car", "Luxury SUV" and "Luxury Car" could be replaced by "Luxury"*

In [209]:
vehicle_changes = {
    "Sports Car": "Luxury",
    "Luxury SUV": "Luxury",
    "Luxury Car": "Luxury"
}

df['Vehicle Class'] = df['Vehicle Class'].replace(vehicle_changes)

In [210]:
df.columns = (df.columns.str.strip().str.replace('_', ' ').str.title())

***Exercise 3: Formatting data types

The data types of many columns in the dataset appear to be incorrect. This could impact the accuracy of our analysis. To ensure accurate analysis, we need to correct the data types of these columns. Please update the data types of the columns as appropriate.

It is important to note that this exercise does not involve handling null values (NaN or None).

*Hint*:
- *Customer lifetime value should be numeric*
- *Number of open complaints has an incorrect format. Look at the different values it takes with `unique()` and take the middle value. As an example, 1/5/00 should be 5. Number of open complaints is a string - remember you can use `split()` to deal with it and take the number you need. Finally, since it should be numeric, cast the column to be in its proper type.*

In [213]:
df.head()

,Customer,State,Gender,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number Of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,1/0/00,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,1/0/00,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,1/0/00,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,1/0/00,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,1/0/00,Personal Auto,Four-Door Car,17.269323


In [214]:
df['Customer Lifetime Value'] = pd.to_numeric(df['Customer Lifetime Value'], errors='coerce')

In [215]:
df.dtypes

Customer                      object
State                         object
Gender                        object
Education                     object
Customer Lifetime Value      float64
Income                       float64
Monthly Premium Auto         float64
Number Of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

In [217]:
df['Number Of Open Complaints'] = df['Number Of Open Complaints'].str.split('/').str[1]

In [218]:
df["Number Of Open Complaints"].value_counts()

Number Of Open Complaints
0    830
1    138
2     50
3     34
4     13
5      6
Name: count, dtype: int64

In [219]:
df.head()

,Customer,State,Gender,Education,Customer Lifetime Value,Income,Monthly Premium Auto,Number Of Open Complaints,Policy Type,Vehicle Class,Total Claim Amount
0,RB50392,Washington,NaN,Master,NaN,0.0,1000.0,0,Personal Auto,Four-Door Car,2.704934
1,QZ44356,Arizona,F,Bachelor,697953.59,0.0,94.0,0,Personal Auto,Four-Door Car,1131.464935
2,AI49188,Nevada,F,Bachelor,1288743.17,48767.0,108.0,0,Personal Auto,Two-Door Car,566.472247
3,WW63253,California,M,Bachelor,764586.18,0.0,106.0,0,Corporate Auto,SUV,529.881344
4,GA49547,Washington,M,High School or Below,536307.65,36357.0,68.0,0,Personal Auto,Four-Door Car,17.269323


## Exercise 4: Dealing with Null values

Identify any columns with null or missing values. Identify how many null values each column has. You can use the `isnull()` function in pandas to find columns with null values.

Decide on a strategy for handling the null values. There are several options, including:

- Drop the rows or columns with null values
- Fill the null values with a specific value (such as the column mean or median for numerical variables, and mode for categorical variables)
- Fill the null values with the previous or next value in the column
- Fill the null values based on a more complex algorithm or model (note: we haven't covered this yet)

Implement your chosen strategy to handle the null values. You can use the `fillna()` function in pandas to fill null values or `dropna()` function to drop null values.

Verify that your strategy has successfully handled the null values. You can use the `isnull()` function again to check if there are still null values in the dataset.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling null values.

After formatting data types, as a last step, convert all the numeric variables to integers.

In [220]:
df.isna().sum()

Customer                     2937
State                        2937
Gender                       3054
Education                    2937
Customer Lifetime Value      2940
Income                       2937
Monthly Premium Auto         2937
Number Of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount           2937
dtype: int64

In [221]:
df.describe()

,Customer Lifetime Value,Income,Monthly Premium Auto,Total Claim Amount
count,1.068000e+03,1071.000000,1071.000000,1071.000000
mean,7.936903e+05,39295.701214,193.234360,404.986909
std,6.434784e+05,30469.427060,1601.190369,293.027260
min,2.004351e+05,0.000000,61.000000,0.382107
25%,4.034080e+05,14072.000000,68.000000,202.157702
50%,5.881742e+05,36234.000000,83.000000,354.729129
75%,8.962872e+05,64631.000000,109.500000,532.800000
max,5.816655e+06,99960.000000,35354.000000,2893.239678


In [222]:
df = df.fillna(df.mean(numeric_only=True))

In [223]:
df.isna().sum()

Customer                     2937
State                        2937
Gender                       3054
Education                    2937
Customer Lifetime Value         0
Income                          0
Monthly Premium Auto            0
Number Of Open Complaints    2937
Policy Type                  2937
Vehicle Class                2937
Total Claim Amount              0
dtype: int64

In [224]:
df.dropna(subset=['Gender'], inplace=True)

In [225]:
df.isna().sum()

Customer                     0
State                        0
Gender                       0
Education                    0
Customer Lifetime Value      0
Income                       0
Monthly Premium Auto         0
Number Of Open Complaints    0
Policy Type                  0
Vehicle Class                0
Total Claim Amount           0
dtype: int64

I filled the null values of all the numerical columns with the mean of each column. This approach is the most accurate the type of data that we are dealing. The categorical columns where drop to avoid create not accurate data. 

## Exercise 5: Dealing with duplicates

Use the `.duplicated()` method to identify any duplicate rows in the dataframe.

Decide on a strategy for handling the duplicates. Options include:
- Dropping all duplicate rows
- Keeping only the first occurrence of each duplicated row
- Keeping only the last occurrence of each duplicated row
- Dropping duplicates based on a subset of columns
- Dropping duplicates based on a specific column

Implement your chosen strategy using the `drop_duplicates()` function.

Verify that your strategy has successfully handled the duplicates by checking for duplicates again using `.duplicated()`.

Remember to document your process and explain your reasoning for choosing a particular strategy for handling duplicates.

Save the cleaned dataset to a new CSV file.

*Hint*: *after dropping duplicates, reset the index to ensure consistency*.

In [227]:
df.duplicated().sum()

np.int64(0)

# Bonus: Challenge 2: creating functions on a separate `py` file

Put all the data cleaning and formatting steps into functions, and create a main function that performs all the cleaning and formatting.

Write these functions in separate .py file(s). By putting these steps into functions, we can make the code more modular and easier to maintain.

*Hint: autoreload module is a utility module in Python that allows you to automatically reload modules in the current session when changes are made to the source code. This can be useful in situations where you are actively developing code and want to see the effects of changes you make without having to constantly restart the Python interpreter or Jupyter Notebook kernel.*

In [ ]:
gender = ["Femenine", "Masculine"]

In [ ]:
import Functions_cleaning
Functions_cleaning.standarized_gender_f(gender)

['Femenine', 'Masculine']

# Bonus: Challenge 3: Analyzing Clean and Formated Data

You have been tasked with analyzing the data to identify potential areas for improving customer retention and profitability. Your goal is to identify customers with a high policy claim amount and a low customer lifetime value.

In the Pandas Lab, we only looked at high policy claim amounts because we couldn't look into low customer lifetime values. If we had tried to work with that column, we wouldn't have been able to because customer lifetime value wasn't clean and in its proper format. So after cleaning and formatting the data, let's get some more interesting insights!

Instructions:

- Review the statistics again for total claim amount and customer lifetime value to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount and a low customer lifetime value. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount, and clients with a low customer lifetime value to be those in the bottom 25% of the customer lifetime value. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile and a customer lifetime value in the bottom 25th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount and low customer lifetime value data. To do so, select both columns of the dataframe simultaneously and pass it to the `.describe()` method. This will give you descriptive statistics, such as mean, median, standard deviation, minimum and maximum values for both columns at the same time, allowing you to compare and analyze their characteristics.

In [229]:
df.describe()
pd.options.display.float_format = '{:.0f}'.format
df.describe()

,Customer Lifetime Value,Income,Monthly Premium Auto,Total Claim Amount
count,954,954,954,954
mean,794200,39576,205,404
std,625222,30717,1696,299
min,200435,0,61,0
25%,411859,13365,68,195
50%,588835,36642,83,350
75%,903293,64948,110,533
max,4479547,99960,35354,2893


In [241]:
claim_75 = df['Total Claim Amount'].quantile(0.75)
clv_25 = df['Customer Lifetime Value'].quantile(0.25)

filtered_df = df[(df['Total Claim Amount'] > claim_75) & (df['Customer Lifetime Value'] < clv_25)]


In [240]:
stats = filtered_df[['Total Claim Amount', 'Customer Lifetime Value']].agg([
    'mean',
    'median',
    'std',
    'min',
    'max', 
    "count"])
print(stats)

        Total Claim Amount  Customer Lifetime Value
mean                   712                   343010
median                 679                   366738
std                    145                    57541
min                    538                   228760
max                   1186                   407913
count                   35                       35


We have 35 customers with higher Total Claim Amount (>75 percentile) and lower Customer Lifetime Value (<25 percentile). These customers have a mean of 712 Total Claim Amount and 343010 Customer Lifetime Value arriving to a maximum of 407913. The Customer Lifetime Value is very low compare to the mean of 794200. Those are customer that pay very low Lifetime Values and recover more than the rest of the customers with a mean of 404. 
